test/multiple-user-principal-types.test.js

```diff
@@ -0,0 +1,978 @@
+      username: 'realm1:foo100',
+      password: 'pass100',
+    };
+
+    var credentialRealmInEmail = {
+      email: 'realm1:foo100@bar.com',
+      password: 'pass100',
+    };
+
+    // create 1 user in each realm / user models with exact same username/email
+    // 4 users in total: realm1/usermodel1, realm1/usermodel2, realm2/usermodel1, realm2/usermodel2
+    var user1a = null;
+    var user1b = null;
+    beforeEach(function(done) {
+      User.create(realm1User, function(err, u1a) {
+        if (err) return done(err);
+        user1a = u1a;
+
+        User2.create(realm1User, function(err, u1b) {
+          if (err) return done(err);
+          user1b = u1b;
+
+          User.create(realm2User, function(err, u1b) {
+            if (err) return done(err);
+
+            User2.create(realm2User, done);
+          });
+        });
+      });
+    });
+
+    it('rejects a user by with bad pass', function(done) {
+      User.login(credentialWithBadPass, function(err, accessToken) {
+        assert(err);
+        assert.equal(err.code, 'LOGIN_FAILED');
+
+        done();
+      });
+    });
+
+    it('logs in a user from user model 1 with realm', function(done) {
+      User.login(credentialWithRealm, function(err, accessToken) {
+        assertGoodToken(accessToken, user1a);
+
+        done();
+      });
+    });
+
+    it('logs in a user from user model 2 with realm', function(done) {
+      User2.login(credentialWithRealm, function(err, accessToken) {
+        assertGoodToken(accessToken, user1b);
+
+        done();
+      });
+    });
+
+    it('logs in a user by username with realm', function(done) {
+      User.login(credentialRealmInUsername, function(err, accessToken) {
+        assertGoodToken(accessToken, user1a);
+
+        done();
+      });
+    });
+
+    it('logs in a user by email with realm', function(done) {
+      User.login(credentialRealmInEmail, function(err, accessToken) {
+        assertGoodToken(accessToken, user1a);
+
+        done();
+      });
+    });
+
+    describe('User.login with realmRequired but no realmDelimiter', function() {
+      beforeEach(function() {
+        User.settings.realmDelimiter = undefined;
+      });
+
+      afterEach(function() {
+        User.settings.realmDelimiter = ':';
+      });
+
+      it('logs in a user by with realm', function(done) {
+        User.login(credentialWithRealm, function(err, accessToken) {
+          assertGoodToken(accessToken, user1a);
+
+          done();
+        });
+      });
+
+      it('rejects a user by with realm in email if realmDelimiter is not set',
+        function(done) {
+          User.login(credentialRealmInEmail, function(err, accessToken) {
+            assert(err);
+            assert.equal(err.code, 'REALM_REQUIRED');
+
+            done();
+          });
+        });
+    });
+  });
+
+  describe('User.logout', function() {
+    it('Logout a user by providing the current accessToken id (using node)', function(done) {
```

Comment:
No need to test both callback an promise variants - pick the Promise variant only and remove ` - promise variant` from the test name.